File: mobile/ios/GuessToday/Features/Game/RulesSheet.swift
import SwiftUI

struct RulesSheet: View {
    let mode: GameMode
    @Environment(\.dismiss) private var dismiss
    @State private var appeared = false

    var body: some View {
        ZStack(alignment: .bottom) {
            Theme.background.ignoresSafeArea()

            ScrollView {
                VStack(spacing: 0) {
                    // Hero header
                    ZStack {
                        // Gradient background
                        LinearGradient(
                            colors: [modeColor.opacity(0.28), modeColor.opacity(0.06), Theme.background],
                            startPoint: .top,
                            endPoint: .bottom
                        )
                        .frame(height: 220)

                        VStack(spacing: Theme.spacing12) {
                            // Icon ring
                            ZStack {
                                Circle()
                                    .fill(modeColor.opacity(0.15))
                                    .frame(width: 90, height: 90)
                                Circle()
                                    .stroke(modeColor.opacity(0.30), lineWidth: 1.5)
                                    .frame(width: 90, height: 90)
                                Image(systemName: mode.iconFilled)
                                    .font(.system(size: 38, weight: .medium))
                                    .foregroundColor(modeColor)
                            }
                            .scaleEffect(appeared ? 1 : 0.55)
                            .opacity(appeared ? 1 : 0)

                            VStack(spacing: 4) {
                                Text("Comment jouer")
                                    .font(Theme.inter(size: 11, weight: .semibold))
                                    .foregroundColor(modeColor.opacity(0.8))
                                    .textCase(.uppercase)
                                    .tracking(1.5)
                                Text(mode.title)
                                    .font(Theme.fraunces(size: 28))
                                    .fontWeight(.bold)
                                    .foregroundColor(Theme.text)
                            }
                            .opacity(appeared ? 1 : 0)
                            .offset(y: appeared ? 0 : 10)
                        }
                        .padding(.top, Theme.spacing24)
                    }

                    // Rules steps
                    VStack(spacing: Theme.spacing8) {
                        ForEach(Array(steps.enumerated()), id: \.offset) { i, step in
                            StepCard(
                                number: i + 1,
                                icon: step.icon,
                                title: step.title,
                                description: step.description,
                                color: modeColor
                            )
                            .opacity(appeared ? 1 : 0)
                            .offset(y: appeared ? 0 : 16)
                            .animation(.spring(response: 0.45, dampingFraction: 0.8).delay(0.12 + Double(i) * 0.07), value: appeared)
                        }
                    }
                    .padding(.horizontal, Theme.spacing16)
                    .padding(.top, Theme.spacing8)

                    // Hint legend (film/series only)
                    if !mode.isWiki {
                        AttemptLegend(color: modeColor)
                            .padding(.horizontal, Theme.spacing16)
                            .padding(.top, Theme.spacing16)
                            .opacity(appeared ? 1 : 0)
                            .animation(.easeOut.delay(0.45), value: appeared)
                    }

                    // Bottom padding for the fixed button
                    Color.clear.frame(height: 100)
                }
            }

            // Fixed CTA button
            VStack(spacing: 0) {
                LinearGradient(
                    colors: [Theme.background.opacity(0), Theme.background],
                    startPoint: .top, endPoint: .bottom
                )
                .frame(height: 32)

                Button {
                    markSeen()
                    dismiss()
                } label: {
                    Text("C'est parti !")
                        .font(Theme.inter(size: 16, weight: .bold))
                        .foregroundColor(Theme.primaryButtonFg)
                        .frame(maxWidth: .infinity)
                        .frame(height: 52)
                        .background(Theme.goldGradient)
                        .clipShape(RoundedRectangle(cornerRadius: Theme.radiusM))
                }
                .padding(.horizontal, Theme.spacing16)
                .padding(.bottom, Theme.spacing24)
                .background(Theme.background)
            }
        }
        .presentationDetents([.large])
        .presentationDragIndicator(.visible)
        .presentationBackground(Theme.background)
        .onAppear {
            withAnimation(.spring(response: 0.5, dampingFraction: 0.72).delay(0.05)) {
                appeared = true
            }
        }
    }

    // MARK: - Steps

    private struct Step {
        let icon: String
        let title: String
        let description: String
    }

    private var steps: [Step] {
        switch mode {
        case .film:
            return [
                Step(icon: "photo.fill", title: "Scène révélée", description: "Une scène du film est affichée. Sers-t'en comme indice visuel pour identifier le film."),
                Step(icon: "list.bullet", title: "Indices progressifs", description: "Jusqu'à 3 indices sont révélés : année de sortie, réalisateur, acteur principal."),
                Step(icon: "film", title: "Proposez un titre", description: "Saisissez votre réponse dans le champ prévu. Vous avez \(maxAttempts) tentatives."),
                Step(icon: "clock.fill", title: "Un défi par jour", description: "Un nouveau film mystère chaque jour à minuit, heure de Paris. Revenez demain !"),
            ]
        case .series:
            return [
                Step(icon: "photo.fill", title: "Scène révélée", description: "Une scène de la série est affichée. Sers-t'en comme indice visuel pour identifier la série."),
                Step(icon: "list.bullet", title: "Indices progressifs", description: "Jusqu'à 3 indices sont révélés : année de sortie, créateur, acteur principal."),
                Step(icon: "tv", title: "Proposez un titre", description: "Saisissez votre réponse dans le champ prévu. Vous avez \(maxAttempts) tentatives."),
                Step(icon: "clock.fill", title: "Un défi par jour", description: "Une nouvelle série mystère chaque jour à minuit, heure de Paris. Revenez demain !"),
            ]
        case .wiki:
            return [
                Step(icon: "eye.slash.fill", title: "Photo masquée", description: "La photo de la personnalité reste cachée tout au long de la partie — découvrez-la à la fin."),
                Step(icon: "list.bullet", title: "Indices progressifs", description: "Des indices sur sa carrière, ses fonctions, ses clubs ou sa vie sont révélés après chaque mauvaise réponse."),
                Step(icon: "person.bust", title: "Proposez un nom", description: "Saisissez votre réponse dans le champ prévu. Vous avez \(maxAttempts) tentatives."),
                Step(icon: "clock.fill", title: "Un défi par jour", description: "Une nouvelle personnalité mystère chaque jour à minuit, heure de Paris. Revenez demain !"),
            ]
        }
    }

    private let maxAttempts = 5

    private var modeColor: Color { mode.color }

    private func markSeen() {
        UserDefaults.standard.set(true, forKey: "rules_seen_\(mode.statsKey)")
    }
}

private extension GameMode {
    var isWiki: Bool { self == .wiki }
}

// MARK: - Step card

private struct StepCard: View {
    let number: Int
    let icon: String
    let title: String
    let description: String
    let color: Color

    var body: some View {
        HStack(alignment: .top, spacing: Theme.spacing12) {
            // Number + icon badge
            ZStack {
                RoundedRectangle(cornerRadius: 10)
                    .fill(color.opacity(0.12))
                    .frame(width: 44, height: 44)
                Image(systemName: icon)
                    .font(.system(size: 18, weight: .medium))
                    .foregroundColor(color)
            }

            VStack(alignment: .leading, spacing: 3) {
                HStack(spacing: 6) {
                    Text("\(number)")
                        .font(Theme.inter(size: 10, weight: .bold))
                        .foregroundColor(color)
                        .frame(width: 16, height: 16)
                        .background(color.opacity(0.15))
                        .clipShape(Circle())
                    Text(title)
                        .font(Theme.inter(size: 14, weight: .semibold))
                        .foregroundColor(Theme.text)
                }
                Text(description)
                    .font(Theme.inter(size: 13))
                    .foregroundColor(Theme.textDim)
                    .lineSpacing(2)
                    .fixedSize(horizontal: false, vertical: true)
            }

            Spacer(minLength: 0)
        }
        .padding(Theme.spacing12)
        .background(Theme.surface)
        .cornerRadius(Theme.radiusM)
        .overlay(
            RoundedRectangle(cornerRadius: Theme.radiusM)
                .stroke(Theme.border, lineWidth: 1)
        )
    }
}

// MARK: - Attempt legend

private struct AttemptLegend: View {
    let color: Color

    var body: some View {
        VStack(alignment: .leading, spacing: Theme.spacing12) {
            Text("Résultat de chaque tentative")
                .font(Theme.inter(size: 11, weight: .semibold))
                .foregroundColor(Theme.textDim)
                .textCase(.uppercase)
                .tracking(1)

            HStack(spacing: 0) {
                LegendDot(color: Theme.green,  label: "Bonne\nréponse")
                Spacer()
                LegendDot(color: Theme.red,    label: "Mauvaise\nréponse")
                Spacer()
                LegendDot(color: Theme.muted,  label: "Tour\npassé", hollow: true)
            }
        }
        .padding(Theme.spacing16)
        .background(Theme.surface)
        .cornerRadius(Theme.radiusM)
        .overlay(
            RoundedRectangle(cornerRadius: Theme.radiusM)
                .stroke(Theme.border, lineWidth: 1)
        )
    }
}

private struct LegendDot: View {
    let color: Color
    let label: String
    var hollow: Bool = false

    var body: some View {
        VStack(spacing: 6) {
            if hollow {
                Circle()
                    .stroke(color, lineWidth: 2)
                    .frame(width: 20, height: 20)
            } else {
                Circle()
                    .fill(color)
                    .frame(width: 20, height: 20)
            }
            Text(label)
                .font(Theme.inter(size: 11))
                .foregroundColor(Theme.textDim)
                .multilineTextAlignment(.center)
                .lineSpacing(1)
        }
    }
}

File: mobile/android/app/src/main/kotlin/fr/guesstoday/features/game/RulesSheet.kt
package fr.guesstoday.features.game

import android.content.Context
import androidx.compose.animation.*
import androidx.compose.animation.core.*
import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.*
import androidx.compose.ui.draw.*
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.*
import androidx.compose.ui.text.font.*
import androidx.compose.ui.unit.*
import fr.guesstoday.ui.theme.AppColors

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RulesSheet(mode: GameMode, onDismiss: () -> Unit) {
    val ctx = LocalContext.current
    var appeared by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) { appeared = true }

    ModalBottomSheet(
        onDismissRequest = {
            markRulesSeen(ctx, mode)
            onDismiss()
        },
        containerColor = AppColors.background,
        dragHandle = { BottomSheetDefaults.DragHandle(color = AppColors.border) }
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp)
                .padding(bottom = 32.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            // Mode icon
            val scale by animateFloatAsState(
                targetValue  = if (appeared) 1f else 0.5f,
                animationSpec = spring(dampingRatio = 0.65f, stiffness = Spring.StiffnessMediumLow),
                label        = "icon_scale"
            )
            Box(
                modifier = Modifier
                    .scale(scale)
                    .alpha(if (appeared) 1f else 0f)
                    .padding(top = 8.dp, bottom = 12.dp)
                    .size(80.dp)
                    .clip(CircleShape)
                    .background(modeColor(mode).copy(alpha = 0.12f)),
                contentAlignment = Alignment.Center
            ) {
                Icon(modeIcon(mode), contentDescription = null, tint = modeColor(mode), modifier = Modifier.size(36.dp))
            }

            // Title
            AnimatedVisibility(visible = appeared, enter = fadeIn() + slideInVertically { it / 3 }) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(
                        "COMMENT JOUER",
                        color    = AppColors.textDim,
                        fontSize = 11.sp,
                        fontWeight = FontWeight.SemiBold,
                        letterSpacing = 1.2.sp,
                    )
                    Spacer(Modifier.height(4.dp))
                    Text(
                        mode.label,
                        color      = AppColors.text,
                        fontSize   = 24.sp,
                        fontFamily = FontFamily.Serif,
                        fontWeight = FontWeight.Bold,
                    )
                }
            }

            Spacer(Modifier.height(20.dp))

            // Rules
            val rules = modeRules(mode)
            rules.forEachIndexed { i, rule ->
                val delay = 100 + i * 70
                var visible by remember { mutableStateOf(false) }
                LaunchedEffect(appeared) {
                    if (appeared) {
                        kotlinx.coroutines.delay(delay.toLong())
                        visible = true
                    }
                }
                AnimatedVisibility(visible = visible, enter = fadeIn() + slideInVertically { it / 2 }) {
                    RuleRow(number = i + 1, text = rule, modifier = Modifier.padding(bottom = 8.dp))
                }
            }

            // Hint legend (film / series only)
            if (!mode.isWiki) {
                Spacer(Modifier.height(4.dp))
                HintLegend()
                Spacer(Modifier.height(16.dp))
            }

            Spacer(Modifier.height(8.dp))

            // CTA
            Button(
                onClick = {
                    markRulesSeen(ctx, mode)
                    onDismiss()
                },
                modifier = Modifier.fillMaxWidth(),
                colors   = ButtonDefaults.buttonColors(containerColor = AppColors.gold, contentColor = Color(0xFF1A0F00)),
                shape    = RoundedCornerShape(10.dp),
            ) {
                Text("C'est parti !", fontWeight = FontWeight.Bold, fontSize = 15.sp)
            }
        }
    }
}

fun rulesAlreadySeen(ctx: Context, mode: GameMode): Boolean {
    val prefs = ctx.getSharedPreferences("game_stats", Context.MODE_PRIVATE)
    return prefs.getBoolean("rules_seen_${mode.statsKey}", false)
}

private fun markRulesSeen(ctx: Context, mode: GameMode) {
    ctx.getSharedPreferences("game_stats", Context.MODE_PRIVATE)
        .edit().putBoolean("rules_seen_${mode.statsKey}", true).apply()
}

@Composable
private fun RuleRow(number: Int, text: String, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(10.dp))
            .background(AppColors.surface)
            .border(1.dp, AppColors.border, RoundedCornerShape(10.dp))
            .padding(12.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.Top,
    ) {
        Box(
            modifier = Modifier
                .size(24.dp)
                .clip(CircleShape)
                .background(AppColors.gold.copy(alpha = 0.12f)),
            contentAlignment = Alignment.Center
        ) {
            Text("$number", color = AppColors.gold, fontSize = 12.sp, fontWeight = FontWeight.Bold, fontFamily = FontFamily.Monospace)
        }
        Text(text, color = AppColors.text, fontSize = 14.sp, lineHeight = 20.sp, modifier = Modifier.weight(1f))
    }
}

@Composable
private fun HintLegend() {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(10.dp))
            .background(AppColors.surface)
            .border(1.dp, AppColors.border, RoundedCornerShape(10.dp))
            .padding(12.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Text(
            "LÉGENDE DES TENTATIVES",
            color    = AppColors.textDim,
            fontSize = 10.sp,
            fontWeight = FontWeight.SemiBold,
            letterSpacing = 1.sp,
        )
        Row(horizontalArrangement = Arrangement.spacedBy(16.dp)) {
            LegendItem(color = AppColors.green, symbol = "●", label = "Bonne réponse")
            LegendItem(color = AppColors.red,   symbol = "●", label = "Mauvaise réponse")
            LegendItem(color = AppColors.muted, symbol = "○", label = "Tour passé")
        }
    }
}

@Composable
private fun LegendItem(color: Color, symbol: String, label: String) {
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(4.dp)) {
        Text(symbol, color = color, fontSize = 10.sp)
        Text(label, color = AppColors.textDim, fontSize = 11.sp)
    }
}

private fun modeIcon(mode: GameMode): ImageVector = when (mode) {
    GameMode.FILM -> Icons.Default.Movie
    GameMode.WIKI -> Icons.Default.AccountBalance
    else          -> Icons.Default.Tv
}

private fun modeColor(mode: GameMode): Color = when (mode) {
    GameMode.FILM -> AppColors.modeFilm
    GameMode.WIKI -> AppColors.modeWiki
    else          -> Color(0xFF6B7CFF)
}

private fun modeRules(mode: GameMode): List<String> = when (mode) {
    GameMode.FILM -> listOf(
        "Devinez le film mystère du jour.",
        "L'image est floue au départ — elle se dévoile progressivement à chaque mauvaise réponse.",
        "Jusqu'à 3 indices sont révélés : année, réalisateur, acteur principal…",
        "Vous avez 5 tentatives. Utilisez la barre de recherche pour proposer un titre.",
        "Un nouveau défi chaque jour à minuit (heure de Paris) !",
    )
    GameMode.WIKI -> listOf(
        "Devinez la personnalité mystère du jour.",
        "Sa photo reste masquée jusqu'à la fin de la partie.",
        "Des indices sur sa carrière, ses fonctions ou ses clubs sont révélés progressivement.",
        "Vous avez 5 tentatives. Utilisez la barre de recherche pour proposer un nom.",
        "Un nouveau défi chaque jour à minuit (heure de Paris) !",
    )
    else -> listOf(
        "Devinez la série mystère du jour.",
        "L'image est floue au départ — elle se dévoile progressivement à chaque mauvaise réponse.",
        "Jusqu'à 3 indices sont révélés : année, créateur, acteur principal…",
        "Vous avez 5 tentatives. Utilisez la barre de recherche pour proposer un titre.",
        "Un nouveau défi chaque jour à minuit (heure de Paris) !",
    )
}

private val GameMode.isWiki: Boolean get() = this == GameMode.WIKI
